### PR TITLE
[JENKINS-27624][JENKINS-42709] Bump Jenkins baseline to Java 8 minimum

### PR DESCRIPTION
**Pull request by batmat**
# Description

Cf. article on jenkins.io: [Jenkins Upgrades To Java 8](https://jenkins.io/blog/2017/01/17/Jenkins-is-upgrading-to-Java-8/)

See also [JENKINS-42709](https://issues.jenkins-ci.org/browse/JENKINS-42709).

Details: 

* [X] Fix generics issues
* [X] Fix infradna/bridge-method-injector to work with Jenkins@JDK8 
* [x] Fix compilation error (in test module)

```
[INFO] 100 warnings
[INFO] -------------------------------------------------------------
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /home/tiste/dev/JENKINS/jenkins/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
```
I guess some help from @kohsuke may be useful here. cc @jglick & @stephenc too.

### Changelog entries

Proposed changelog entries:

* JENKINS-42709: Make Jenkins buildable with JDK 8
* ...

### Submitter checklist

- [X] JIRA issue is well described
- [x] Link to JIRA ticket in description, if appropriate
- [x] Appropriate autotests or explanation to why this change has no tests
- [X] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

### Desired reviewers

@mention

### CODE REVIEW

**Comment by jglick** [comment]
```java
        // checks the class version
        if (readShort(off + 6) > Opcodes.V1_7) {
            throw new IllegalArgumentException();
        }
```

Did you try just upgrading `asm-debug-all` from 4.2, say to 5.2?

```java
        // checks the class version
        if (readShort(off + 6) > Opcodes.V1_8) {
            throw new IllegalArgumentException();
        }
```

**Comment by batmat** [comment]
@jglick Pretty sure I did try that, among other combinations. But I will retry to make sure.

**Comment by batmat** [comment]
To sum up, the hint of Jesse above pushed me into the right direction. I think I had touched Jenkins' pom.xml instead of the plugin's when twiddling with it.

So, now I'm past it, the good news is that the WAR build succeeds (though I didn't smoke test it yet), only the `test` module fails because of a Javac error. I think this is coming from the inference fixes that have been added while introducing lambdas, and now the code is ambiguous. Looking into it.

https://github.com/jenkinsci/jenkins/blob/749346896abbdce8e074d593ab37175f59a84c29/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java#L89 does not compile, and shows:
```
[ERROR] /home/tiste/dev/JENKINS/jenkins/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
[INFO] 1 error
```

**Comment by batmat** [comment]
So, now it fails with the following only in CI. And not locally with `1.8.0_121`.
I can reproduce using `1.8.0_66`, so I suspect it's a javac bug that got fixed between the two.

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/jenkins/util/JenkinsJVMRealTest.java:[23,42] error: unreported exception T; must be caught or declared to be thrown
[INFO] 1 error
```

**Comment by batmat** [comment]
@jenkinsci/code-reviewers review welcome: this PR bumps Jenkins core to JDK8 minimum.

**Comment by aheritier** [comment]
Maybe https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core should be removed ?

File: pom.xml
Comment (jglick) [inline pom.xml]: 1.15 has been released acc. to @kohsuke.

File: core/src/main/java/hudson/model/User.java
Comment (jglick) [inline core/src/main/java/hudson/model/User.java]: Might suffice to just replace `Run<?,?>` with `Run` (untested). Or to add a qualifier to `filter` like

```java
return RunList.fromJobs(Jenkins.getInstance().allItems(Job.class)).<Something<Here>>filter(new Predicate<Run<?,?>>() {
```

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java]: Ditto.

**Review by jglick — APPROVED** [review]

File: test/src/test/java/jenkins/util/JenkinsJVMRealTest.java
Comment (oleg-nenashev) [inline test/src/test/java/jenkins/util/JenkinsJVMRealTest.java]: Hm? Why would you need such change?

Comment (batmat) [inline test/src/test/java/jenkins/util/JenkinsJVMRealTest.java]: @oleg-nenashev I tried to separate each "fix" in a dedicated commit. This is https://github.com/jenkinsci/jenkins/pull/2802/commits/77dfa64b6a173b58c7f61107dcc51a61f7a3466a

Comment (batmat) [inline test/src/test/java/jenkins/util/JenkinsJVMRealTest.java]: The build logs exhibiting the issue: https://ci.jenkins.io/job/Core/job/jenkins/job/PR-2802/2/execution/node/28/log/

Copying the end for future ref when the build has been wiped out:

```
proprietary API and may be removed in a future release
21:18:06 [INFO] 100 warnings 
21:18:06 [INFO] -------------------------------------------------------------
21:18:06 [INFO] -------------------------------------------------------------
21:18:06 [ERROR] COMPILATION ERROR : 
21:18:06 [INFO] -------------------------------------------------------------
21:18:06 [ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/jenkins/util/JenkinsJVMRealTest.java:[23,42] error: unreported exception T; must be caught or declared to be thrown
21:18:06 [ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
21:18:06 [INFO] 2 errors 
21:18:06 [INFO] -------------------------------------------------------------
21:18:06 [INFO] ------------------------------------------------------------------------
21:18:06 [INFO] Reactor Summary:
21:18:06 [INFO] 
21:18:06 [INFO] Jenkins main module ................................ SUCCESS [ 45.970 s]
21:18:06 [INFO] Jenkins cli ........................................ SUCCESS [ 42.146 s]
21:18:06 [INFO] Jenkins core ....................................... SUCCESS [03:12 min]
21:18:06 [INFO] Jenkins war ........................................ SUCCESS [01:23 min]
21:18:06 [INFO] Tests for Jenkins core ............................. FAILURE [ 56.812 s]
21:18:06 [INFO] ------------------------------------------------------------------------
21:18:06 [INFO] BUILD FAILURE
21:18:06 [INFO] ------------------------------------------------------------------------
21:18:06 [INFO] Total time: 07:18 min
21:18:06 [INFO] Finished at: 2017-03-29T21:18:05+00:00
21:18:06 [INFO] Final Memory: 100M/594M
21:18:06 [INFO] ------------------------------------------------------------------------
21:18:06 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.0:testCompile (default-testCompile) on project test: Compilation failure: Compilation failure:
21:18:06 [ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/jenkins/util/JenkinsJVMRealTest.java:[23,42] error: unreported exception T; must be caught or declared to be thrown
21:18:06 [ERROR] /home/jenkins/workspace/Core_jenkins_PR-2802-7MEOUWG2NJTLSSLTVDVDIXOKU5BJ5UJ5PCL2QYQPQXAHRMFCDSMA/test/src/test/java/hudson/model/GetEnvironmentOutsideBuildTest.java:[89,30] error: no suitable method found for buildAndAssertSuccess(AbstractProject)
21:18:06 [ERROR] -> [Help 1]
21:18:06 [ERROR] 
21:18:06 [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
21:18:06 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
21:18:06 [ERROR] 
21:18:06 [ERROR] For more information about the errors and possible solutions, please read the following articles:
21:18:06 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
21:18:06 [ERROR] 
21:18:06 [ERROR] After correcting the problems, you can resume the build with the command
21:18:06 [ERROR]   mvn <goals> -rf :test
```

File: pom.xml
Comment (batmat) [inline pom.xml]: Fixed with https://github.com/jenkinsci/jenkins/pull/2802/commits/496d574b8c31f06b3c005cbdc84370bb0d47c69a

File: core/src/main/java/hudson/model/User.java
Comment (batmat) [inline core/src/main/java/hudson/model/User.java]: I had tried quite a few combinations before finding that one, for instance with `Run` only:

image

So I'm ready to change to something else, but I thought the fix I have was readable/simple enough anyway (?).

**Comment by batmat** [comment]
@olivergondza I would really appreciate your review, especially with your LTS Officer cap, as this will possibly propagate to next one (or next-next) if we merge it soonish as initially planned.

File: test/src/test/java/jenkins/util/JenkinsJVMRealTest.java
Comment (batmat) [inline test/src/test/java/jenkins/util/JenkinsJVMRealTest.java]: @oleg-nenashev does this address your question/concern? Thanks

Comment (oleg-nenashev) [inline test/src/test/java/jenkins/util/JenkinsJVMRealTest.java]: it does

**Review by oleg-nenashev — APPROVED** [review]
I am fine with the PR itself, but maybe it makes sense to postpone it till the next weekly. #2826 is a serious regression, and I would rather allow Java 7 users to pick it. Not a blocker, just a quite biased IMHO

**Comment by daniel-beck** [comment]
Could we get rid of occurrences of `MaxPermSize` then? It means nothing on JDK 8 but spams the logs.

**Comment by batmat** [comment]
@daniel-beck done. Arnaud had already asked for it, and I said I would do it in a followup PR (to keep that PR a bit _shorter_). As you were now two to request it, I went ahead and did it here.

**Comment by batmat** [comment]
I believe all concerns have been addressed and that this pull request is ready to merge.

If nobody objects in the meantime, or does it before me, I plan to merge this PR before the end of the week. 

**Which means that the next weekly will then require Java 8 to run.**

**Review by szpak — APPROVED** [review]
LGTM

**Review by olivergondza — APPROVED** [review]

**Review by jglick — APPROVED** [review]

**Comment by jglick** [comment]
> #2826 is a serious regression, and I would rather allow Java 7 users to pick it.

@oleg-nenashev perhaps you missed our policy update when getting rid of the `rc` branch, but in summary: if you find a serious regression and fix it, you can ask @kohsuke to cut an out-of-order trunk release. No need to wait a week.

**Comment by oleg-nenashev** [comment]
@jglick Asn a requester of several out-of-order releases on this year, I definitely know about this opportunity. And I also know about obstacles like KK's reachability. In the case of this fix it has landed too late (late Friday) for making out-of-order release meaningful. 

BTW I think we should just go forward with this PR

**Comment by oleg-nenashev** [comment]
Merging without squash since there are 2 authors && the change is not a backporting candidate

**Comment by oleg-nenashev** [comment]
Uh oh, the merge button dissapeared :D

**Comment by batmat** [comment]
Thank you everyone! \o/

**Pull request merged**
